### PR TITLE
Upgrade to react-router v4

### PR DESCRIPTION
**Pull request by kxmbrian**
Main Changes
- To change location, inject `history` as a prop (e.g. using `withRouter`) and call `push` on it instead of calling `push` on the imported `browserHistory`.
- URL parameters are now found in the `match` prop instead of `params` directly.
- As [recommended](https://reacttraining.com/react-router/web/guides/redux-integration), actions that change the browser location are passed a `history` object.
- Add helpers to help with test setup. Removed `muiTheme` global since it now not used anywhere.
- As [recommended](https://reacttraining.com/react-router/web/guides/testing), use `MemoryRouter` to set up router context instead of manually stubbing it out.

### CODE REVIEW

**Comment by allenwq** [comment]
@kxmbrian Is there any way to redirect without injecting`withRouter` ?   Now there're more boilerplates when creating a new component > <

**Comment by kxmbrian** [comment]
@allenwq [This answer](http://stackoverflow.com/a/42124328/2817855) gives a good summary of the ways to redirect. 

It might be possible to use `<Redirect>`, but it is triggered by a `render`. Perhaps one way to use it is to do it in the same style as the notifications snackbar - we can dispatch a redux action with the destination route, store it, render `Redirect`, then clear the store after the redirect. Doesn't seem very natural or clean though.

Do you have any suggestions?

**Comment by allenwq** [comment]
@kxmbrian Yah I think using `Recdirect` component won't make things better,  how about we maintain a global history somewhere and import it when needed ?

**Comment by kxmbrian** [comment]
@allenwq I switched to using a global history and nested the routes to pages for an individual survey under SurveyLayout. Now each of the pages receive `survey` as a prop instead of all the `surveys`.

File: client/app/bundles/course/survey/pages/SurveyIndex/SurveysTable.jsx [hunk 1]
```diff
@@ -115,10 +114,10 @@ class SurveysTable extends React.Component {
                   { survey.allow_response_after_end ? survey.time_bonus_exp : '-' }
                 </TableRowColumn>
                 <TableRowColumn colSpan={5}>
-                  { intl.formatDate(survey.start_at, standardDateFormat) }
+                  { moment(survey.start_at).format('DD-MM-YYYY') }
```
Comment (allenwq) [inline client/app/bundles/course/survey/pages/SurveyIndex/SurveysTable.jsx]: at some point we should standard this as well.  now some places using `DD-MM-YYYY`, and some other places uses `DD MM YY`

File: client/app/__test__/setup.js [hunk 2]
```diff
@@ -12,13 +13,29 @@ const timeZone = "Asia/Singapore";
...
+const buildContextOptions = store => ({
+  context: { intl, store, muiTheme },
+  childContextTypes: {
+    muiTheme: React.PropTypes.object,
+    store: React.PropTypes.object,
+    intl: intlShape,
+  },
+});
+
+const buildRouterProps = params => ({
```
Comment (allenwq) [inline client/app/__test__/setup.js]: Is this still necessary ?

**Comment by allenwq** [comment]
@kxmbrian the global history looks good !  

But it would be nice if you can reorganize the commits 😄  Because the PR is big and it's a bit hard to review commit by commit also since things added in the earlier commits are removed later. 

@weiqingtoh do you have any comments? I know that you were working on this earlier.

**Comment by kxmbrian** [comment]
The commits have been squashed. Will fix the date formatting in a housekeeping PR.

**Comment by weiqingtoh** [comment]
I can't comment much since I was still trying to figure out React-router, but it seems to make sense based on what I can vaguely understand. Thanks @kxmbrian 👍

File: client/app/bundles/course/survey/pages/ResponseShow/__test__/index.test.jsx [hunk 3]
```diff
@@ -18,10 +13,11 @@ describe('<ResponseShow />', () => {
...
 
     // Mount response show page and wait for data to load
-    Object.defineProperty(window.location, 'pathname', { value: responseUrl });
     mount(
-      <ResponseShow params={{ courseId, surveyId, responseId }} />,
-      generateContextOptions(storeCreator({}))
+      <MemoryRouter initialEntries={[responseUrl]}>
+        <ResponseShow {...{ survey: {}, courseId, surveyId, match: { params: { responseId } } }} />
+      </MemoryRouter>,
+      buildContextOptions(storeCreator({}))
```
Comment (allenwq) [inline client/app/bundles/course/survey/pages/ResponseShow/__test__/index.test.jsx]: @kxmbrian possible to make `MemoryRouter` a context option?

**Comment by allenwq** [comment]
LGTM otherwise !

File: client/app/bundles/course/survey/pages/ResponseShow/__test__/index.test.jsx (hunk 3 above)
Comment (kxmbrian) [inline client/app/bundles/course/survey/pages/ResponseShow/__test__/index.test.jsx]: After nesting the pages in `SurveyLayout`, only two of the tests use `MemoryRouter`. Just removed the other occurrences.

Comment (allenwq) [inline client/app/bundles/course/survey/pages/ResponseShow/__test__/index.test.jsx]: OK, we leave it for now.

**Pull request merged**